### PR TITLE
Refactor for OOP

### PR DESCRIPTION
**Pull request by aidanmelen**
- Major refactored for class inheritance. Most notably, this improves on the `Bobcat` because now there is a dedicated class for the `BobcatAPI` which returns the response `JSON`. The `BobcatAPI` is subclass to `Bobcat` which will be responsible for parsing the `JSON`.

- Added method (`BobcatConnection().search()`) to scan the local network for the Bobcat miner ip address. 

- Overhauled log design with better Terminal coloring and Discord Embed messages.

- I took the fall for science and applied **real** actions against my **real** bobcat and recorded the resulting responses and state changes from the Bobcat. I documented and redacted in the `docs/observations` directory. I can now use this data to implement a rock solid Bobcat Autopilot. 🚀🚀🚀

### CODE REVIEW

**Comment by manjot96** [comment]
Could you add a OTA check to make sure the bobcat's version is up to date ?

**Comment by manjot96** [comment]
Any way to make the bobcat class more generic so one could add support for other miners ? like nebra and sense cap

**Comment by aidanmelen** [comment]
> Could you add a OTA check to make sure the bobcat's version is up to date ?

An OTA check would be nice but it would ultimately be non-actionable by the autopilot routine since users don't control when their bobcat gets the OTA. Would this be a warning? What would be the threshold? 1 day out-of-date? 5 days? I don't know if there is a sane choice.

That being said, I have wanted to log an event when the bobcat receives the latest OTA. I feel like this would be useful because Bobcats often encounter issues due to bad OTA rollouts. This would simply show up as an info log.

**Comment by manjot96** [comment]
I've been researching a bit and I'd love to setup your sdk on balena cloud with Docker so its on a raspberry pi but remote as well

**Comment by aidanmelen** [comment]
> Any way to make the bobcat class more generic so one could add support for other miners ? like nebra and sense cap

Unfortunately I don't know anything about those manufacturers. I only own Bobcat. Therefore I don't have the data required nor the plans to build automation for anything other than Bobcat.

**Review by manjot96 — APPROVED** [review]
LGTM

**Review by manjot96 — DISMISSED** [review]

**Comment by manjot96** [comment]
> > Any way to make the bobcat class more generic so one could add support for other miners ? like nebra and sense cap
> 
> Unfortunately I don't know anything about those manufacturers. I only own Bobcat. Therefore I don't have the data required nor the plans to build automation for anything other than Bobcat.

I'd love to add support for them as I'm receiving those next week. I can contribute to the code

**Comment by aidanmelen** [comment]
> LGTM

Thanks I will likely merge sometime next week. Still have some outstanding changes.

**Comment by aidanmelen** [comment]
> > > Any way to make the bobcat class more generic so one could add support for other miners ? like nebra and sense cap
> 
> > 
> 
> > Unfortunately I don't know anything about those manufacturers. I only own Bobcat. Therefore I don't have the data required nor the plans to build automation for anything other than Bobcat.
> 
> 
> 
> I'd love to add support for them as I'm receiving those next week. I can contribute to the code

I am sorry but that is out of the scope for this project. But you are welcome to fork the project for anything useful. 😃

**Comment by manjot96** [comment]
> > LGTM
> 
> Thanks I will likely merge sometime next week. Still have some outstanding changes.

yes I noticed that now auto pilot still needs to be filled out. Will be awaiting the update. thanks

File: src/bobcat_miner/autopilot.py [hunk 1]
```diff
@@ -1,582 +1,397 @@
...
 
-    #     if not self.is_syncing():
-    #         self.logger.critical("The autosync failed")
-    #         self.logger.critical("Manual intervention is required")
-    #         self.logger.debug(
-    #             "Contact Bobcat Support: https://bobcatminer.zendesk.com/hc/en-us/articles/4412998083355-Contact-Support"
-    #         )
+class BobcatAutopilot(Bobcat):
+
+    DIAGNOSER_STATES = [
```
Comment (manjot96) [inline src/bobcat_miner/autopilot.py:17]: Should probably move this to a seperate file

**Pull request merged**